lib/httplib/httpheaders.go

```diff
@@ -21,51 +21,77 @@ package httplib
 	"strings"
 	"time"
 
-	"golang.org/x/exp/maps"
-
 	"github.com/gravitational/teleport/api/client/proto"
 )
 
-var defaultContentSecurityPolicy = map[string]string{
-	"default-src": "'self'",
+type cspMap map[string][]string
+
+var defaultContentSecurityPolicy = cspMap{
+	"default-src": {"'self'"},
 	// specify CSP directives not covered by `default-src`
-	"base-uri":        "'self'",
-	"form-action":     "'self'",
-	"frame-ancestors": "'none'",
+	"base-uri":        {"'self'"},
+	"form-action":     {"'self'"},
+	"frame-ancestors": {"'none'"},
 	// additional default restrictions
-	"object-src": "'none'",
-	"img-src":    "'self' data: blob:",
-	"style-src":  "'self' 'unsafe-inline'",
+	"object-src": {"'none'"},
+	"img-src":    {"'self'", "data:", "blob:"},
+	"style-src":  {"'self'", "'unsafe-inline'"},
 }
 
-var defaultFontSrc = map[string]string{"font-src": "'self' data:"}
+var defaultFontSrc = cspMap{"font-src": {"'self'", "data:"}}
+var defaultConnectSrc = cspMap{"connect-src": {"'self'", "wss:"}}
 
-var stripeSecurityPolicy = map[string]string{
+var stripeSecurityPolicy = cspMap{
 	// auto-pay plans in Cloud use stripe.com to manage billing information
-	"script-src": "'self' https://js.stripe.com",
-	"frame-src":  "https://js.stripe.com",
+	"script-src": {"'self'", "https://js.stripe.com"},
+	"frame-src":  {"https://js.stripe.com"},
+}
+
+var wasmSecurityPolicy = cspMap{
+	"script-src": {"'wasm-unsafe-eval'"},
+}
+
+func deduplicate(elements []string) []string {
+	encountered := map[string]bool{}
+	result := []string{}
+
+	for v := range elements {
+		if encountered[elements[v]] {
+			// Do not add element when encountered before
+		} else {
+			encountered[elements[v]] = true
+			result = append(result, elements[v])
+		}
+	}
+	return result
 }
 
 // combineCSPMaps combines multiple CSP maps into a single map.
-// When multiple of the input cspMaps have the same key, the
-// latter map's value takes precedence.
-func combineCSPMaps(cspMaps ...map[string]string) map[string]string {
-	combined := make(map[string]string)
+// When multiple of the input cspMaps have the same key, their
+// respective lists are concatenated.
+func combineCSPMaps(cspMaps ...cspMap) cspMap {
+	combinedMap := make(cspMap)
+
 	for _, cspMap := range cspMaps {
-		maps.Copy(combined, cspMap)
+		for key, value := range cspMap {
+			combinedMap[key] = append(combinedMap[key], value...)
+			combinedMap[key] = deduplicate(combinedMap[key])
+		}
 	}
-	return combined
+
+	return combinedMap
 }
 
 // getContentSecurityPolicyString combines multiple CSP maps into a single
 // CSP string, alphabetically sorted by the directive key.
-// When multiple of the input cspMaps have the same key, the
-// latter map's value takes precedence.
-func getContentSecurityPolicyString(cspMaps ...map[string]string) string {
+// When multiple of the input cspMaps have the same key, their
+// respective lists are concatenated.
+func getContentSecurityPolicyString(cspMaps ...cspMap) string {
```

Comment:
I wonder if it would be worth benchmarking this and checking whether caching the results makes much of a difference.

With these changes we went from a relatively static CSP to code that is allocating, looping over maps, deduplicating, and building strings on _every single request_.

Maybe it doesn't matter at all, but given that there's probably only a few possible combinations of potential CSPs we will return, precomputing might make sense.

Reply:
I added caching for the index csp (because it was easier, and is the most-often-hit) which significantly improved the mean execution time. I modified the test like so
<details>
<summary>diff</summary>

```diff
diff --git a/lib/httplib/httplib_test.go b/lib/httplib/httplib_test.go
index 5b3c4712ff..7ace3b368d 100644
--- a/lib/httplib/httplib_test.go
+++ b/lib/httplib/httplib_test.go
@@ -26,6 +26,7 @@ import (
 	"net/url"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/gravitational/roundtrip"
 	"github.com/julienschmidt/httprouter"
@@ -347,12 +348,40 @@ func TestSetIndexContentSecurityPolicy(t *testing.T) {
 	} {
 		t.Run(tt.name, func(t *testing.T) {
 			h := make(http.Header)
-			SetIndexContentSecurityPolicy(h, tt.features, tt.urlPath)
+
+			var times []time.Duration
+			for i := 0; i < 1000; i++ {
+				start := time.Now()
+				SetIndexContentSecurityPolicy(h, tt.features, tt.urlPath)
+				elapsed := time.Since(start)
+				times = append(times, elapsed)
+			}
+
 			actualCsp := h.Get("Content-Security-Policy")
 			for k, v := range tt.expectedCspVals {
 				expectedCspSubString := fmt.Sprintf("%s %s;", k, v)
 				require.Contains(t, actualCsp, expectedCspSubString)
 			}
+
+			// Calculate min, max, and mean
+			var min, max, total time.Duration
+			min = times[0]
+			max = times[0]
+			for _, t := range times {
+				if t < min {
+					min = t
+				}
+				if t > max {
+					max = t
+				}
+				total += t
+			}
+			mean := total / time.Duration(len(times))
+
+			// Print min, max, and mean
+			t.Logf("Min time: %v", min)
+			t.Logf("Max time: %v", max)
+			t.Logf("Mean time: %v", mean)
 		})
 	}
 }
```

</details>

and got the following results

<details>
<summary>pre-cache</summary>

```
$ /opt/homebrew/bin/go test -test.v -timeout 30s -tags desktop_access_rdp -run ^TestSetIndexContentSecurityPolicy$ github.com/gravitational/teleport/lib/httplib
=== RUN   TestSetIndexContentSecurityPolicy
=== PAUSE TestSetIndexContentSecurityPolicy
=== CONT  TestSetIndexContentSecurityPolicy
=== RUN   TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm)
    httplib_test.go:382: Min time: 4.583µs
    httplib_test.go:383: Max time: 362.083µs
    httplib_test.go:384: Mean time: 6.644µs
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm)
    httplib_test.go:382: Min time: 5.708µs
    httplib_test.go:383: Max time: 127.834µs
    httplib_test.go:384: Mean time: 7.206µs
=== RUN   TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm)
    httplib_test.go:382: Min time: 5.625µs
    httplib_test.go:383: Max time: 123.041µs
    httplib_test.go:384: Mean time: 8.443µs
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm)
    httplib_test.go:382: Min time: 7.584µs
    httplib_test.go:383: Max time: 166.084µs
    httplib_test.go:384: Mean time: 9.786µs
--- PASS: TestSetIndexContentSecurityPolicy (0.03s)
    --- PASS: TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm) (0.01s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm) (0.01s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm) (0.01s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm) (0.01s)
PASS
ok  	github.com/gravitational/teleport/lib/httplib	0.482s
```

</details>

<details>
<summary>with-cache</summary>

```
$ /opt/homebrew/bin/go test -test.v -timeout 30s -tags desktop_access_rdp -run ^TestSetIndexContentSecurityPolicy$ github.com/gravitational/teleport/lib/httplib
=== RUN   TestSetIndexContentSecurityPolicy
=== PAUSE TestSetIndexContentSecurityPolicy
=== CONT  TestSetIndexContentSecurityPolicy
=== RUN   TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm)
    httplib_test.go:382: Min time: 3.125µs
    httplib_test.go:383: Max time: 405.625µs
    httplib_test.go:384: Mean time: 4.777µs
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm)
    httplib_test.go:382: Min time: 3.833µs
    httplib_test.go:383: Max time: 192.209µs
    httplib_test.go:384: Mean time: 5.49µs
=== RUN   TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm)
    httplib_test.go:382: Min time: 4.083µs
    httplib_test.go:383: Max time: 122.916µs
    httplib_test.go:384: Mean time: 4.95µs
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm)
    httplib_test.go:382: Min time: 4.666µs
    httplib_test.go:383: Max time: 110.458µs
    httplib_test.go:384: Mean time: 5.681µs
--- PASS: TestSetIndexContentSecurityPolicy (0.02s)
    --- PASS: TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm) (0.01s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm) (0.01s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm) (0.01s)
PASS
ok  	github.com/gravitational/teleport/lib/httplib	0.408s
```

</details>

(I also tried a version without Sprintf, but that gave me worse results. I also double checked if the `Max time` was correlated with the first call to the function, but it appeared to be random. I'm not sure what to make of that, but I elected against an `init` function to pre-fill the cache).

I will take a look at adding caching to the other csp functions next.

Reply:
I realize I was being dumb above and failing to cache before the expensive map iterating and looping operation of `getContentSecurityPolicyString`. With de3f5a29ddf8bc1c03b2d8060c03902d3bfd3c2e, 58664290c769b8f021086b93581735317d2276ad, and d9f6ec25860ae266ab911af983ce7b12c68ce1a8 I've added caching for all the variations and get an OOM improvement with the cache like:

```
$ /opt/homebrew/bin/go test -test.v -timeout 30s -tags desktop_access_rdp -run ^TestSetIndexContentSecurityPolicy$ github.com/gravitational/teleport/lib/httplib 2>&1 | tee out.log
=== RUN   TestSetIndexContentSecurityPolicy
=== PAUSE TestSetIndexContentSecurityPolicy
=== CONT  TestSetIndexContentSecurityPolicy
=== RUN   TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm)
    httplib_test.go:382: Min time: 125ns
    httplib_test.go:383: Max time: 10.292µs
    httplib_test.go:384: Mean time: 179ns
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm)
    httplib_test.go:382: Min time: 125ns
    httplib_test.go:383: Max time: 9.833µs
    httplib_test.go:384: Mean time: 182ns
=== RUN   TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm)
    httplib_test.go:382: Min time: 83ns
    httplib_test.go:383: Max time: 10.084µs
    httplib_test.go:384: Mean time: 174ns
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm)
    httplib_test.go:382: Min time: 125ns
    httplib_test.go:383: Max time: 10µs
    httplib_test.go:384: Mean time: 174ns
--- PASS: TestSetIndexContentSecurityPolicy (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm) (0.00s)
PASS
ok  	github.com/gravitational/teleport/lib/httplib	0.421s
```